usd/src/usd_parser/USDLinks.cc

```diff
@@ -0,0 +1,163 @@
+ * limitations under the License.
+ *
+*/
+
+#include "USDLinks.hh"
+
+#include <string>
+
+#pragma push_macro ("__DEPRECATED")
+#undef __DEPRECATED
+#include <pxr/usd/usdGeom/cube.h>
+#include <pxr/usd/usdGeom/cylinder.h>
+#include <pxr/usd/usdGeom/gprim.h>
+#include <pxr/usd/usdGeom/mesh.h>
+#include <pxr/usd/usdGeom/sphere.h>
+#include "pxr/usd/usdPhysics/rigidBodyAPI.h"
+#pragma pop_macro ("__DEPRECATED")
+
+#include <ignition/common/Filesystem.hh>
+#include <ignition/common/Util.hh>
+
+#include <ignition/math/Inertial.hh>
+
+#include "sdf/Geometry.hh"
+#include "sdf/Link.hh"
+
+#include "sdf/usd/usd_parser/USDTransforms.hh"
+
+namespace sdf
+{
+// Inline bracke to help doxygen filtering.
+inline namespace SDF_VERSION_NAMESPACE {
+//
+namespace usd
+{
+void GetInertial(
+  const pxr::UsdPrim &_prim,
+  ignition::math::Inertiald &_inertial)
+{
+  float mass;
+  pxr::GfVec3f centerOfMass;
+  pxr::GfVec3f diagonalInertia;
+
+  ignition::math::MassMatrix3d massMatrix;
+
+  if (_prim.HasAPI<pxr::UsdPhysicsRigidBodyAPI>())
+  {
+    if (pxr::UsdAttribute massAttribute =
+      _prim.GetAttribute(pxr::TfToken("physics:mass")))
+    {
+      massAttribute.Get(&mass);
+
+      if (pxr::UsdAttribute centerOfMassAttribute =
+        _prim.GetAttribute(pxr::TfToken("physics:centerOfMass"))) {
+        centerOfMassAttribute.Get(&centerOfMass);
+
+      }
+      if (pxr::UsdAttribute diagonalInertiaAttribute =
+        _prim.GetAttribute(pxr::TfToken("physics:diagonalInertia"))) {
+        diagonalInertiaAttribute.Get(&diagonalInertia);
+      }
+
+      // Added a diagonal inertia to avoid crash with the physics engine
+      if (diagonalInertia == pxr::GfVec3f(0, 0, 0))
+      {
+        diagonalInertia = pxr::GfVec3f(0.0001, 0.0001, 0.0001);
+      }
+
+      if (mass < 0.0001)
+      {
+        mass = 0.1;
+      }
+
+      massMatrix.SetMass(mass);
+      massMatrix.SetDiagonalMoments(
+        ignition::math::Vector3d(
+          diagonalInertia[0],
+          diagonalInertia[1],
+          diagonalInertia[2]));
+
+      _inertial.SetPose(ignition::math::Pose3d(
+          ignition::math::Vector3d(
+            centerOfMass[0], centerOfMass[1], centerOfMass[2]),
+          ignition::math::Quaterniond(1.0, 0, 0, 0)));
+
+      _inertial.SetMassMatrix(massMatrix);
+    }
+  }
+  else
+  {
+    auto parent = _prim.GetParent();
+    if (parent)
+    {
+      GetInertial(parent, _inertial);
+    }
+  }
+}
+
+sdf::Link * ParseUSDLinks(
+  const pxr::UsdPrim &_prim,
+  const std::string &_nameLink,
+  sdf::Link *_link,
+  USDData &_usdData,
+  int &/*_skipPrim*/)
+{
+  std::string primNameStr = _prim.GetPath().GetName();
+  std::string primPathStr = pxr::TfStringify(_prim.GetPath());
+  std::string primType = _prim.GetPrimTypeInfo().GetTypeName().GetText();
+
+  std::pair<std::string, std::shared_ptr<USDStage>> data =
+    _usdData.FindStage(primNameStr);
+
+  // Is this a new link ?
+  if (_link == nullptr)
+  {
+    _link = new sdf::Link();
```

Comment:
we are getting a raw pointer from `sdf::Link` I was trying to keep it simple

Reply:
I changed the API slightly to use `std::optional<sdf::Link>` instead of raw pointers. I think it's a safer approach since it avoids the need for manually allocating/de-allocating memory